oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/OAuth2ReauthorizeRequest.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.client.web;
+
+import org.springframework.security.core.Authentication;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClient;
+import org.springframework.util.Assert;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Represents a request the {@link OAuth2AuthorizedClientManager} uses to
+ * {@link OAuth2AuthorizedClientManager#reauthorize(OAuth2ReauthorizeRequest) re-authorize}
+ * the provided {@link OAuth2AuthorizedClient#getClientRegistration() client}.
+ *
+ * @author Joe Grandja
+ * @since 5.2
+ * @see OAuth2AuthorizeRequest
+ * @see OAuth2AuthorizedClientManager
+ */
+public class OAuth2ReauthorizeRequest extends OAuth2AuthorizeRequest {
```

Comment:
I don't think it makes sense for `OAuth2ReauthorizeRequest` to extend `OAuth2AuthorizeRequest`. What does it mean for `OAuth2ReauthorizeRequest` to be passed into `OAuth2AuthorizedClientManager.authorize`? The types should enforce correct usage.

Reply:
> I don't think it makes sense for `OAuth2ReauthorizeRequest` to extend `OAuth2AuthorizeRequest`

This makes perfect sense to me. A request to re-authorize, (e.g. token is expired) is considered an authorization request as well. Therefore, a re-authorization request is an authorization request.

> What does it mean for `OAuth2ReauthorizeRequest` to be passed into `OAuth2AuthorizedClientManager.authorize()`

Why would this even happen? The API is clear `authorize(OAuth2AuthorizeRequest)` so even if a `OAuth2ReauthorizeRequest` was passed into `authorize()` it would only be used as an `OAuth2AuthorizeRequest` and **never cast** to an `OAuth2ReauthorizeRequest`.
Can you expand on the use case because as far as I see it the user would not be using the API correctly in this instance if they were to pass in an `OAuth2ReauthorizeRequest`, but the method implementation would be correctly implemented based on the clear definition of the operation and it's input.

Also, the added benefit of having the hierarchy `OAuth2ReauthorizeRequest` extends `OAuth2AuthorizeRequest` is the `DefaultOAuth2AuthorizedClientManager` will only need one `setContextAttributesMapper(Function<OAuth2AuthorizeRequest, Map<String, Object>> contextAttributesMapper)`. It would not be ideal to have a second `setContextAttributesMapper(Function<OAuth2ReauthorizeRequest, Map<String, Object>> contextAttributesMapper)` (typed `OAuth2ReauthorizeRequest`)

Reply:
The API should enforce proper usage. It is surprising behavior that the API allows an
OAuth2AuthorizedClient to be provided to authorize via the OAuth2ReauthorizeRequest but it will be completely ignored.

Reply:
I'd like to hear your answer regarding my question...

> Why would this even happen?...

Is this really a valid scenario? Like I already mentioned this would be a user error on the usage of the API. But the API would still behave as expected, ignoring the `OAuth2ReauthorizeRequest  ` attributes.

Furthermore, based on this understanding does this issue become invalid? See [comment](https://github.com/spring-projects/spring-security/issues/5633#issuecomment-410059242). Because based on this issue, an `OAuth2LoginAuthenticationToken` could get passed into an `OAuth2AuthorizationCodeAuthenticationProvider` which is identically scenario we are discussing here. @jzheaux Let's get your input here as I'm still not convinced.

Reply:
> Is this really a valid scenario?

That is exactly my point. The API should not allow this to happen, but it does.

> But the API would still behave as expected,

As a user I find this confusing. If I create an `OAuth2ReauthorizeRequest` providing the `OAuth2AuthorizedClient`, it is surprising to me that the `OAuth2AuthorizedClient` would be ignored.

Reply:
Is there a need for the two types and the separate methods? Looking at the default implementations, there seems to be a large overlap. If both are indeed authorize requests, as the hierarchy implies, perhaps `OAuth2AuthorizeRequest` could optionally hold an `Oauth2AuthorizedClient`.

Reply:
I personally like the clarity of the API for `OAuth2AuthorizedClientManager` having both `authorize()` and `reauthorize()`. However, I'm open to discussing the removal of `reauthorize()`. If we were to do that than we have this:


```java
public interface OAuth2AuthorizedClientManager {
	@Nullable
	OAuth2AuthorizedClient authorize(OAuth2AuthorizeRequest authorizeRequest);
}

public class OAuth2AuthorizeRequest {
	private final String clientRegistrationId;
	private final OAuth2AuthorizedClient authorizedClient;
	private final Authentication principal;
	private final HttpServletRequest servletRequest;
	private final HttpServletResponse servletResponse;

	public OAuth2AuthorizeRequest(String clientRegistrationId, Authentication principal,
									HttpServletRequest servletRequest, HttpServletResponse servletResponse) {

	}

	public OAuth2AuthorizeRequest(OAuth2AuthorizedClient authorizedClient, Authentication principal,
									HttpServletRequest servletRequest, HttpServletResponse servletResponse) {

	}
	

```

`OAuth2AuthorizeRequest` would provide 2 constructors and would be similar to the members in `OAuth2AuthorizationContext` which is passed downstream to the `OAuth2AuthorizedClientProvider`(s).

@rwinch What are your thoughts on this proposed design?

Reply:
I'm ok with this proposal as it solve my concern and aligns with our previous discussions (i.e. we can have two separate methods with two request types or the same method with a single request type).